addons/binding/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/model/Realtime.java

```diff
@@ -18,35 +16,42 @@
  */
 public class Realtime extends ErrorResponse {
 
-    @SerializedName(value = "current", alternate = "current_ma")
+    private static final int MILLIWATT_TO_WATT = 1000;
+    private static final int MILLIAMP_TO_AMP = 1000;
+    private static final int WATTHOUR_TO_KILOWATTHOUR = 1000;
+    private static final int MILLIVOLT_TO_VOLT = 1000;
+
     private double current;
-    @SerializedName(value = "power", alternate = "power_mw")
     private double power;
-    @SerializedName(value = "total", alternate = "total_wh")
     private double total;
-    @SerializedName(value = "voltage", alternate = "voltage_mv")
     private double voltage;
 
+    // JSON names used for v2 hardware
+    private double currentMa;
+    private double powerMw;
+    private double totalWh;
+    private double voltageMv;
+
     public double getCurrent() {
-        return current;
+        return currentMa > 0.0 ? currentMa / MILLIAMP_TO_AMP : current;
     }
 
     public double getPower() {
-        return power;
+        return powerMw > 0.0 ? powerMw / MILLIWATT_TO_WATT : power;
     }
 
     public double getTotal() {
-        return total;
+        return totalWh > 0.0 ? totalWh / WATTHOUR_TO_KILOWATTHOUR : total;
     }
 
     public double getVoltage() {
-        return voltage;
+        return voltageMv > 0.0 ? voltageMv / MILLIVOLT_TO_VOLT : voltage;
     }
 
     @Override
     public String toString() {
-        return "current:" + current + ", power:" + power + ", total:" + total + ", voltage:" + voltage
```

Comment:
Maybe its is useful for logging purposes if you log whether this is an old or new object, now that is visible?

Reply:
I guess it matters when you want to know the exact content of the json, but there are other log lines that will log the exact json if needed. So I don't think the additional complexity of this toString will make be worth it.

Reply:
I assumed that it could be useful to know whether your current state is based on calculated values or values returned from the external api, but I leave that up to the two of you.

Reply:
I originally tried doing something like this thinking it would help in identifying future issues with new TP-Link emeter JSON.  Long story short, the resulting `toString()` was way more complex than the actual V2 support and ended up making the code fairly messy.
 
Doing so effectively resulted in saying 'using V2 values' in this class.  I then realised this class really wasn't the place identify the hardware version, which can more appropriately and accurately be identified in the `GetSysinfo` class (where it is reported by the device).  Also then noticed the `Realtime.toString()` method isn't directly logged and the TRACE details from `Connection.sendCommand()` with the JSON result were far more useful than `Realtime.toString()` was ever going to be.

So my preference would be to leave it out too.